mcsema/Arch/ABI.cpp

```diff
@@ -390,6 +396,32 @@ static const char *ReturnValVar(llvm::CallingConv::ID cc, llvm::Type *type) {
   }
 }
 
+static llvm::Type* RetrieveArgumentType(llvm::Type *original_type, unsigned index) {
+  if (original_type->isPointerTy()) {
+    return original_type;
+  }
+  // float complex may look as <2xfloat>
+  if (auto seq_type = llvm::dyn_cast<llvm::SequentialType>(original_type)) {
+    return seq_type->getElementType();
+  //long double complex may look as { x86_fp80, x86_fp80 }
+  } else if (auto struct_type = llvm::dyn_cast<llvm::StructType>(original_type)) {
+    return struct_type->getElementType(index);
+  } else {
+    return original_type;
+  }
+}
+
+// llvm::CompositeType as common parent does not provide getNumElements
+static uint64_t GetNumberOfElements(llvm::Type* original_type) {
+  if (auto vec_type = llvm::dyn_cast<llvm::VectorType>(original_type)) {
+    return vec_type->getNumElements();
+  } else if (auto struct_type = llvm::dyn_cast<llvm::StructType>(original_type)) {
+    return struct_type->getNumElements();
+  } else {
```

Comment:
Do array types need to be handled? I mean, I dunno how you even pass an array to a function in C/C++ that doesn't just get turned into a pointer.

I'm sort of curious about things like passing `struct Foo { int x[2]; };` typed objects as arguments, as they're shaped similarly to `struct Bar { int x0; int x1; };` but would be treated as having 1 elements instead of two.

Reply:
The same goes for things like nesting structs.

Reply:
Sequential type is used for vectors `struct Foo{ float a; float b };` will be `<2xfloat>` when passed/returned by value ( with extra pain of being in one register ). It is quite possible, that this only happens to floats, since from `struct Bar { int x0; int x1 }` becomes simple `i64`. But I don't have any confirmation, so code is more general.

Both `Foo` and `Bar` as presented in post above would be simply `i64` when passed/returned  by value from function.
If Bar looked like `struct Bar { int x0; int x1; int x2}` then it would be passed as `{ i64, i32 }`.
Once structure is too big, it simply becomes pointer (if passed by value it gains `byval` attribute).

Reply:
I agree that `<2xfloat>` may be a special case and therefore handled before the loop, simplifying it a bit